### PR TITLE
Meal-sharing: Frontend

### PR DESCRIPTION
**Pull request by Divyajg**
I didn't get the review for the frontend branch.  I merged frontend branch and created new branch frontend-v2.

### CODE REVIEW

File: src/client/components/AddMeal.js [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  const [title, setTitle] = useState("");
+  const [description, setDescription] = useState("");
+  const [location, setLocation] = useState("");
+  const [When, setWhen] = useState("");
+  const [reservations, setReservations] = useState("");
+  const [price, setPrice] = useState("");
+
+  function newMeal() {
+    {
+      if (title !== "") {
```
Comment (islam-fawzy25) [inline src/client/components/AddMeal.js]: Here you are checking if the title filled so can submit new meal and i think you did not mean that 
Sorry i tried to add meal only with title and that is happened but the all-meals page broken now

File: src/client/components/AddMeal.js [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            setInputstate(false);
+            alert("Meal Added");
+          });
+      }
+    }
+  }
+  return (
+    <div>
+      <Header />
+      <div className="inputForm">
```
Comment (islam-fawzy25) [inline src/client/components/AddMeal.js:41]: We need to use form to make sure that all the required inputs is filled  or we can not submit before filled all the inputs 
Check this example it is not the perfect but would help you what i mean 
https://github.com/islam-fawzy25/meal-sharing/blob/master/src/client/components/meals/newMeal/AddNewMeal.js

**Review by islam-fawzy25 — COMMENTED** [review]
Hi Divya,
Great job, I would like to start by suggestion to make this Great job looks perfect:
1- Lest's organise the files and make components/containers folders
FX:
 - Components
  -Meals (folder)
    -Meals.component.js
    -meals.style.css
-Containers or Pages
 -HomePage
  -HomePage.container.js
  -homePage.style.css
Same as in the final project, It is the best practice.

**Comment by miltonhultgren** [comment]
@islam-fawzy25 In modern React with hooks, what is the difference between a component and a container?
Does this best practice still apply or even exist? 

My view is that that kind of split made more sense when we had only access to state in classes which also where components.
I think modern React offers us more options with hooks to be able to separate data from views.

So rather than splitting it by component/container I would split it by components/hooks but keep the hook close to the component it is used with if it is not shared with many components.
I could still argue for having a folder for the page components, but not because they are handling data but because they are the "sub-roots" to your app once you get past the router so they act as entry points. They might set up the context providers for the rest of the components but I think hooks allows us to move the data management closer to where it is used rather than prop drilling.

So for me it has less to do with the complexity of the component or the data, but rather where in the hierarchy they live. If a component or hook is only used in one page it should live under the folder for that page. This is similar to the split by feature structure used in backend apps (rather than split by type, controller/service/ repository).

@Divyajg  So I'd much rather suggest this kind of structure (if you find it worth the time to change this):

- components (for components that are used in more pages)
  - component1
    - JSX file
    - hook file
- hooks (for hooks that are used in more pages)
- pages
  - home
    - components (used only in this page)
      - JSX file
      - hook file
    - hooks (used only in this page)
    - JSX file for the Page

**Comment by miltonhultgren** [comment]
I have been thinking about this all day! And I realized that I did not address the main goal of Islam's comment! 
I should have asked "why was it best practice to split by component/container". 

TL;DR: If you aim to create your code in smaller pieces that know as little as possible (single responsibility) it'll be easier to maintain your code, and your folder structure should reflect this. 

Learning from the purpose of a best practice makes it so that we can apply the wisdom even if technology changes. In my view, the main goal of that pattern is the idea "single responsibility". 

The main idea is that, the less things each part of your code knows about the better. Because the less things it knows about the less reason it has to change. And less reason for code to change means less work which also means less chances to put bugs into the code. 
If one change in your system requires that a lot of different pieces change then these pieces know too much about each other, and thus they likely have more than one responsibility.

So think of a simple React component that only accepts props and does some rendering. It knows very little, it's only responsible for rendering the data it receives to the screen, and maybe reacting to events by calling a callback it got from props. It knows nothing else, it only has to change if a) the rendering should change or b) the specific data it gets changed. For all other changes in the code base, it does not need to change. 
A container should be the opposite, it knows only about how to get the data and which component to pass that data to. It only needs to change if a) the data to fetch changes or b) there is another component to use for rendering. 

So while we used to think of Pages as containers, there is actually two parts of a page. The data loading for that page, and the layout rendering for that page. So it would be better if that was two components, one that loads the data and one that only does the layout. 
And this is exactly what a hook is. A hook only handles data, it doesn't know about rendering. So you still gain the value of the single responsibility idea.
A similar example, in your App you have all of the routes but it also contains other code. It would make sense to extract all of the routes to another component called AppRouter. That component only knows about how to define routes, and App only knows that there is some general layout and an AppRouter.

Related to this topic is the idea of "worker code" and "coordination code". If you imagine a recipe, it contains a set of steps to perform to achieve a tasty outcome. Each step doesn't really know about the other steps, and the steps could be combined in other ways to achieve a different (maybe) tasty outcome. So the thing that decides the outcome is the recipe itself, the thing that coordinates the steps.
Code works in a similar way. You have functions that do "work" and then you have functions that only call other functions in the right order to get something done. And that's because knowing how to do "work" is one responsibility, and knowing how to organize "work" is a different responsibility! And you can apply this on the function level, the component level (page layout vs product widget) and even on the system level.
I work at Elastic and our main product Elasticsearch is a distributed database, so you have many nodes and each node has a different responsibility. So we have some data storage nodes and some search nodes. But we also have a node type that we call a coordinating node, it it's only job is to tell the other nodes how to do their jobs! 

So to summarize, if you aim to create your code in smaller pieces that know as little as possible it'll be easier to maintain your code, and your folder structure should reflect this.

File: public/index.html [hunk 3]
```diff
@@ -1,38 +1,28 @@
...
+    <link rel="shortcut icon" href="Logo.png" />
+    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no" />
     <!-- Primary Meta Tags -->
     <title>Meal Sharing App</title>
     <meta name="title" content="Meal Sharing App" />
-    <meta
-      name="description"
-      content="Website api for sharing meals together. Invite your best friends or your family for dinner!"
-    />
+    <meta name="description" content="Website api for sharing meals together. Invite your best friends or your family for dinner!" />
```
Comment (miltonhultgren) [inline public/index.html:12]: A little detail but is it a website or an API? :D 
It's either a website (that happens to use an API)
Or it's an API (that can be consumed by a website, or a phone app or something else)

File: src/client/App.js [hunk 4]
```diff
@@ -5,28 +5,39 @@ import Meals from "./components/Meals";
...
+import AllMeals from "./components/All-meals";
 import { ContextProvider } from "./components/DataContext";
 import "./App.css";
 
 function App() {
   return (
     <Router className="app">
       <Switch>
-        <Route exact path="/home" component={HomePage} />
+        <Route exact path="/" component={HomePage} />
```
Comment (miltonhultgren) [inline src/client/App.js:21]: One thing that stands out to me is how this is the only component here that is called Page, I would expect to see AboutPage and MealsPage etc. since they're all pages.

File: src/client/App.js [hunk 5]
```diff
@@ -5,28 +5,39 @@ import Meals from "./components/Meals";
...
+        <Route exact path="/meals/:id" component={MealById} />
+        <Route exact path="/reviews/:id" component={ReviewById} />
+        <Route exact path="/searchMeal" component={AutoSearch} />
         <ContextProvider>
           <Meals />
+          <AllMeals />
+          <Reviews />
+          <MealById />
+          <Reservations />
         </ContextProvider>
```
Comment (miltonhultgren) [inline src/client/App.js:38]: I'm confused why this is here.

These are all components that seem to be pages, why are they not defined as a `<Route>`?
Is it because you needed access to the data context? 
You most likely want to wrap the data context around the `<Switch>` instead if you want to to be available for all children of your app.

Like this, these components won't be rendered the right way.

File: src/client/components/About.js [hunk 6]
```diff
@@ -1,24 +1,135 @@
...
-      <h2 id="personalInfo"><a>Do Not Sell My Personal Information</a></h2>
-      <h2 id="cookie"><a>Cookie Settings</a></h2>
-      <h2 id="contact"><a>Contact US</a></h2>
-      <h2 id="location"><a>Location</a></h2>
-      
-      </>
+    <>
+      <Header />
+      <div>
+        <div className="flex2">
```
Comment (miltonhultgren) [inline src/client/components/About.js:9]: flex2 seems like a poor class name, can you think of a name that instead describes the effect of the class?

File: src/client/components/About.js [hunk 7]
```diff
@@ -1,24 +1,135 @@
...
+          <div className="mealBorder">
```
Comment (miltonhultgren) [inline src/client/components/About.js:10]: Meal border seems out of place in component that doesn't talk about meals?

File: src/client/components/About.js [hunk 8]
```diff
@@ -1,24 +1,135 @@
...
+              Our History, Values In Action, Leadership Team, Investor
+              Relations, News andp//#endregion Notifications, Franchise Info.
+            </p>
+          </div>
+          <div className="mealBorder">
+            <h2 id="careers">
+              <a>Careers</a>
+            </h2>
+            <p>Education, Opportunities, Employee Perks, Working With Us.</p>
+          </div>
```
Comment (miltonhultgren) [inline src/client/components/About.js:24]: This pattern seems to repeat itself, that is usually a sign that there is a component that can be extracted here.

File: src/client/components/AddMeal.js [hunk 9]
```diff
@@ -0,0 +1,126 @@
+import React, { useState } from "react";
+import Header from "./Header";
+import Footer from "./Footer";
+
+export default function AddMeal() {
+  const [inputState, setInputstate] = useState(false);
+  const [title, setTitle] = useState("");
+  const [description, setDescription] = useState("");
+  const [location, setLocation] = useState("");
+  const [When, setWhen] = useState("");
```
Comment (miltonhultgren) [inline src/client/components/AddMeal.js:10]: Nitpick(naming): Prefer camelCase for variables and functions, save PascalCase for classes and types.

File: src/client/components/AddMeal.js [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+            location: location,
+            When_date: When,
+            max_reservations: reservations,
+            price: price,
+            created_date: new Date(),
+          }),
+        })
+          .catch((e) => {
+            setError(e);
+            alert("please enter the Valid details");
```
Comment (miltonhultgren) [inline src/client/components/AddMeal.js]: I'd expect setting the error would render an error prompt in the page, so no need to alert. Browser alerts are very disrupting so I would avoid them! :)

File: src/client/components/AddMeal.js [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+  }
+  return (
+    <div>
+      <Header />
+      <div className="inputForm">
+        <label>
+          Title:{" "}
+          <input
+            onChange={(e) => setTitle(e.target.value)}
+            className="addMargin"
```
Comment (miltonhultgren) [inline src/client/components/AddMeal.js]: Try to name class names so they define the effect but not as a verb. "addMargin" sounds like a function name. Instead declare what it will be, now how it will do it "margin-small" for example.

File: src/client/components/AddMeal.js [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+            type="text"
+            defaultValue={""}
+            placeholder="Title"
+            required
+          />{" "}
```
Comment (miltonhultgren) [inline src/client/components/AddMeal.js]: I think you want to add CSS margin/padding instead of filling your code with {" "}.
Remember, HTML is about page structure, not layout or visual design, leave that to CSS.

File: src/client/components/AddMeal.js [hunk 13]
```diff
@@ -0,0 +1,126 @@
+import React, { useState } from "react";
+import Header from "./Header";
+import Footer from "./Footer";
+
+export default function AddMeal() {
+  const [inputState, setInputstate] = useState(false);
```
Comment (miltonhultgren) [inline src/client/components/AddMeal.js:6]: Ah, I guess this is something like "isSending", but you never use it to disable the submit button.

File: src/client/components/AddReview.js [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+import DataContext from "./DataContext";
+
+export default function AddReview() {
+  const { upcomingMeals } = useContext(DataContext);
+  const [inputState, setInputstate] = useState(false);
+  const [title, setTitle] = useState("");
+  const [description, setDescription] = useState("");
+  const [mealId, setMealId] = useState("");
+  const [stars, setStars] = useState("");
+  function newReview() {
```
Comment (miltonhultgren) [inline src/client/components/AddReview.js:13]: Try to name functions as verbs or action. "newReview" is a noun, how about "addNewReview"?

File: src/client/components/AddReview.js [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+    {
+      if (title !== "") {
```
Comment (miltonhultgren) [inline src/client/components/AddReview.js]: You could remove `setInputstate` and this if, instead you could place it on the submit button like:
`<button onClick={newReview} disabled={title !== ""}>Submit</button>`

File: src/client/components/All-meals.js [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+                    {meal.title}{" "}
+                  </li>
+                  <li>{meal.description}</li>
+                  <li>{meal.price} DKK</li>
+                  <li>Event on: {meal.when_date.slice(0, 10)}</li>
+                </div>
+              );
+            })
+          )}
+        </ol>
```
Comment (miltonhultgren) [inline src/client/components/All-meals.js:34]: This looks like a component that could live on their own, "MealsList"?

The React philosophy is really to create very, very small components so don't be scared to extract things!

File: src/client/components/AutoSearch.js [hunk 17]
```diff
@@ -0,0 +1,51 @@
+import React, { useState, useEffect } from "react";
+import Header from "./Header";
+import Footer from "./Footer";
+import { Link } from "react-router-dom";
+
+const AutoSearch = () => {
+    const [data, setData]=useState([])
```
Comment (miltonhultgren) [inline src/client/components/AutoSearch.js:7]: This is my personal trigger, everything in software is data! 
What is special about this data? Put that into the name. It seems like searchResults?

File: src/client/components/AutoSearch.js [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+                <span className="boldFont"> Meal Price: </span>{meal.price} <br />
+                <Link to={`/meals/${meal.id}`}>
+                <button>
+                  Explore this meal 🥗
+                </button> </Link>
+              </li>
+            );
+          })
+        )}
+      </ol>
```
Comment (miltonhultgren) [inline src/client/components/AutoSearch.js:44]: This code looks very familiar, this looks like "MealsList"!

File: src/client/components/HomePage.js [hunk 19]
```diff
@@ -62,28 +31,18 @@ export default function HomePage(props) {
       </section>
 
       <div className="aboutPage">
-        <ul>
-          <li>
-            <a href="meal-sharing/about/">about us</a>
+        <ul className="listBlock">
```
Comment (miltonhultgren) [inline src/client/components/HomePage.js:34]: ListBlock also seems to be a component that can be extracted!

File: src/client/components/MealById.js [hunk 20]
```diff
@@ -0,0 +1,158 @@
+import React, { useContext, useState } from "react";
+import Header from "./Header";
+import Footer from "./Footer";
+import { useParams } from "react-router-dom";
+import DataContext from "./DataContext";
+import { Link } from "react-router-dom";
+
+export default function MealById() {
```
Comment (miltonhultgren) [inline src/client/components/MealById.js:8]: I'd be wary to name components based on the API they are using! What if the API endpoint name changes or the component gets data from something else?

Instead, name it for what it shows the user!

File: src/client/components/MealById.js [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+          }),
+        })
+          .catch((e) => {
+            setError(e);
+            alert("please enter the Valid details");
+          })
+          .finally(() => {
+            setInputstate(false);
+            alert("Your Reservation has been saved. Enjoy the Experience.");
+          });
```
Comment (miltonhultgren) [inline src/client/components/MealById.js:39]: This pattern is also repeated in every page component, maybe it would make sense to extract a function or a hook that can handle this?

File: src/client/components/MealById.js [hunk 22]
```diff
@@ -0,0 +1,158 @@
...
+      }
+    }
+  }
+  const mealById = newMeals.find((meal) => meal.id === parseInt(id));
```
Comment (miltonhultgren) [inline src/client/components/MealById.js:43]: Just a little detail, on each render, for each items in newMeals, you will call parseInt.
Have a look at using `useMemo` at the start of this component instead!

File: src/client/components/MealById.js [hunk 23]
```diff
@@ -0,0 +1,158 @@
...
+    } else if (parseInt(stars) === 2) {
+      return <span>⭐⭐</span>;
+    } else if (parseInt(stars) === 3) {
+      return <span>⭐⭐⭐</span>;
+    } else if (parseInt(stars) === 4) {
+      return <span>⭐⭐⭐⭐</span>;
+    } else if (parseInt(stars) === 5) {
+      return <span>⭐⭐⭐⭐⭐</span>;
+    }
+  }
```
Comment (miltonhultgren) [inline src/client/components/MealById.js:60]:
```suggestion
  function reviewStars(stars) {
    const starsToRender = Array(stars).fill(0); // Creates an array with as many items as stars, and puts a 0 in each slot so we can iterate over it
    return (
      <span>
        {starsToRender.map(() => ⭐)}
      </span>
    );
  }
```

With this code you can support basically any number of stars. Code that includes that many if branches is usually a sign to re-think your data model!

File: src/client/components/MealById.js [hunk 24]
```diff
@@ -0,0 +1,158 @@
...
+  return (
+    <div>
+      <Header />
+      <div>
+        {!mealById ? (
```
Comment (miltonhultgren) [inline src/client/components/MealById.js:65]: When your component has multiple ternaries in it's markup, it's a good hint you should extract a component.

File: src/client/components/Reservations.js [hunk 25]
```diff
@@ -1,5 +1,119 @@
...
+  if (error) return <p>{error}</p>;
+
+  const [inputState, setInputstate] = useState(false);
+  const [name, setName] = useState("");
+  const [phoneNumber, setPhoneNumber] = useState("");
+  const [numberOfGuests, setNumberOfGuests] = useState("");
+  const [email, setEmail] = useState("");
+  const [mealId, setMealId] = useState("");
+
+  function newReservation() {
```
Comment (miltonhultgren) [inline src/client/components/Reservations.js:19]: Also feel free to extract this function to it's own file. I don't need to read the details of this function to understand the component and vice versa. This just gives the readers more details which they might not be looking for at that time.

It's like reading an article in a newspaper right? They first give you a big headline, to decide if you care at all (the function name). Then they give you a quick summary (this function calls these 3 other functions with good names), and if you still care you keep reading and the article slowly reveals more and more details.
When you feel like you understand enough or found what you were looking for, you stop. You can choose not to read the rest which is even more details. Code should slowly reveal itself the same way, from the big ideas down to the details.

File: src/client/components/Reviews.js [hunk 26]
```diff
@@ -1,5 +1,56 @@
...
+import DataContext from "./DataContext";
+import { Link } from "react-router-dom";
 
 export default function Reviews() {
-  return <h2>Reviews</h2>;
+  const { meals, reviews, loading, error } = useContext(DataContext);
+  if (loading) return <h3>Loading...</h3>;
+  if (error) return <p>{error}</p>;
+
+  function reviewStars(stars){
```
Comment (miltonhultgren) [inline src/client/components/Reviews.js:12]: When you start extracting code into their own modules, it becomes easier to remember what code you already have modules for so you don't have to repeat yourself! Less code = less bugs.

File: src/client/components/componentStyle.css [hunk 27]
```diff
@@ -3,51 +3,114 @@
     padding: 10px;
```
Comment (miltonhultgren) [inline src/client/components/componentStyle.css:3]: I tend to not ever review the actual CSS used, if it looks good the CSS is usually good. 
What I do look at is the names of classes, like are you writing btn instead of button.
Or are you naming classes contentDiv because it's added to a `<div>` or naming it flexContainer simply because the class uses flex. 

CSS classes should describe what a part of HTML is, so writing submitButton is fine. It's not because it uses <button> but because it looks like a button to the user! And when they click it, it submis the form.

Other than that, I would avoid "joined selector" styles, like "submitButton ul", instead name what that `<ul>` means in the context of the submitButton and create a dedicated class for that and add it to the ul directly.

File: src/client/components/DataContext.js [hunk 28]
```diff
@@ -4,26 +4,70 @@ const DataContext = React.createContext();
```
Comment (miltonhultgren) [inline src/client/components/DataContext.js:4]: I don't really understand why this context exists.
In some components you pull data from this context, in others they pull their own data with fetch.

If the aim is to fetch the same data less often there are better ways to solve that.
If the concern is hitting the database too much, put a small cache in the backend.
If the concern is hitting the backend too much, put a small cache in the frontend.

Here is a good article: https://betterprogramming.pub/exploring-caching-techniques-in-react-d30bbb78d54d

In general, I would not worry about that in this kind of app, just call the API on each page for the data you need.

File: src/client/components/DataContext.js [hunk 29]
```diff
@@ -4,26 +4,70 @@ const DataContext = React.createContext();
...
+      meal.availableSeats = meal.max_reservations-meal.reservations;
+    }
+    return meal;
+  }) 
+
+  const upcomingMeals=newMeals.filter((meal) => Date.parse(meal.when_date) >= Date.parse(new Date()));
 
   return (
-    <DataContext.Provider value={{ meals, loading, error }}>
+    <DataContext.Provider
```
Comment (miltonhultgren) [inline src/client/components/DataContext.js:59]: I'm also very bothered by the fact that all of this unrelated data is placed into the same provider, if anything I would split this into one provider for each type of data.

**Review by miltonhultgren — COMMENTED** [review]
I have left you a ton of comments but overall this looks fairly alright. 
I have two main points of feedback:

1. It's worth using tools like VSCode's built in code [formatter](https://code.visualstudio.com/docs/editor/codebasics#_formatting). Specially if you want to show it to possible employers. You want each and every file to be formatted in the same predictable way. Presenting code that is easy to read is key when trying to impress.  So start by just going over every file, use the auto format and commit that.

2. Many components and files are too big, try to be more eager with [extracting things](https://medium.com/swlh/how-to-write-great-react-c4f23f2f3f4f). In general, small things are easier to read and understand. That ties into point 1. By splitting things apart you make it easier to understand, and quick to read which means whoever is looking at your code will both see more of your code and understand more of your code in the same time because each file will be easy to grasp. The discussion above about folder structure also plays into this. How can you structure all these small files in a way that it is easy for me to find what I'm curious about. I would love to easily find a file called "router.jsx" that tells me all of the pages your app has. Then a folder called "pages" so I can pick which page I want to look at. Inside each page folder, I'd expect to find a file called "XPage.jsx" where I can see quickly what data this page pulls in and which other components it renders with that data.

**Comment by islam-fawzy25** [comment]
> I have been thinking about this all day! And I realized that I did not address the main goal of Islam's comment! I should have asked "why was it best practice to split by component/container".
> 
> TL;DR: If you aim to create your code in smaller pieces that know as little as possible (single responsibility) it'll be easier to maintain your code, and your folder structure should reflect this.
> 
> Learning from the purpose of a best practice makes it so that we can apply the wisdom even if technology changes. In my view, the main goal of that pattern is the idea "single responsibility".
> 
> The main idea is that, the less things each part of your code knows about the better. Because the less things it knows about the less reason it has to change. And less reason for code to change means less work which also means less chances to put bugs into the code. If one change in your system requires that a lot of different pieces change then these pieces know too much about each other, and thus they likely have more than one responsibility.
> 
> So think of a simple React component that only accepts props and does some rendering. It knows very little, it's only responsible for rendering the data it receives to the screen, and maybe reacting to events by calling a callback it got from props. It knows nothing else, it only has to change if a) the rendering should change or b) the specific data it gets changed. For all other changes in the code base, it does not need to change. A container should be the opposite, it knows only about how to get the data and which component to pass that data to. It only needs to change if a) the data to fetch changes or b) there is another component to use for rendering.
> 
> So while we used to think of Pages as containers, there is actually two parts of a page. The data loading for that page, and the layout rendering for that page. So it would be better if that was two components, one that loads the data and one that only does the layout. And this is exactly what a hook is. A hook only handles data, it doesn't know about rendering. So you still gain the value of the single responsibility idea. A similar example, in your App you have all of the routes but it also contains other code. It would make sense to extract all of the routes to another component called AppRouter. That component only knows about how to define routes, and App only knows that there is some general layout and an AppRouter.
> 
> Related to this topic is the idea of "worker code" and "coordination code". If you imagine a recipe, it contains a set of steps to perform to achieve a tasty outcome. Each step doesn't really know about the other steps, and the steps could be combined in other ways to achieve a different (maybe) tasty outcome. So the thing that decides the outcome is the recipe itself, the thing that coordinates the steps. Code works in a similar way. You have functions that do "work" and then you have functions that only call other functions in the right order to get something done. And that's because knowing how to do "work" is one responsibility, and knowing how to organize "work" is a different responsibility! And you can apply this on the function level, the component level (page layout vs product widget) and even on the system level. I work at Elastic and our main product Elasticsearch is a distributed database, so you have many nodes and each node has a different responsibility. So we have some data storage nodes and some search nodes. But we also have a node type that we call a coordinating node, it it's only job is to tell the other nodes how to do their jobs!
> 
> So to summarize, if you aim to create your code in smaller pieces that know as little as possible it'll be easier to maintain your code, and your folder structure should reflect this.



> I have been thinking about this all day! And I realized that I did not address the main goal of Islam's comment! I should have asked "why was it best practice to split by component/container".
> 
> TL;DR: If you aim to create your code in smaller pieces that know as little as possible (single responsibility) it'll be easier to maintain your code, and your folder structure should reflect this.
> 
> Learning from the purpose of a best practice makes it so that we can apply the wisdom even if technology changes. In my view, the main goal of that pattern is the idea "single responsibility".
> 
> The main idea is that, the less things each part of your code knows about the better. Because the less things it knows about the less reason it has to change. And less reason for code to change means less work which also means less chances to put bugs into the code. If one change in your system requires that a lot of different pieces change then these pieces know too much about each other, and thus they likely have more than one responsibility.
> 
> So think of a simple React component that only accepts props and does some rendering. It knows very little, it's only responsible for rendering the data it receives to the screen, and maybe reacting to events by calling a callback it got from props. It knows nothing else, it only has to change if a) the rendering should change or b) the specific data it gets changed. For all other changes in the code base, it does not need to change. A container should be the opposite, it knows only about how to get the data and which component to pass that data to. It only needs to change if a) the data to fetch changes or b) there is another component to use for rendering.
> 
> So while we used to think of Pages as containers, there is actually two parts of a page. The data loading for that page, and the layout rendering for that page. So it would be better if that was two components, one that loads the data and one that only does the layout. And this is exactly what a hook is. A hook only handles data, it doesn't know about rendering. So you still gain the value of the single responsibility idea. A similar example, in your App you have all of the routes but it also contains other code. It would make sense to extract all of the routes to another component called AppRouter. That component only knows about how to define routes, and App only knows that there is some general layout and an AppRouter.
> 
> Related to this topic is the idea of "worker code" and "coordination code". If you imagine a recipe, it contains a set of steps to perform to achieve a tasty outcome. Each step doesn't really know about the other steps, and the steps could be combined in other ways to achieve a different (maybe) tasty outcome. So the thing that decides the outcome is the recipe itself, the thing that coordinates the steps. Code works in a similar way. You have functions that do "work" and then you have functions that only call other functions in the right order to get something done. And that's because knowing how to do "work" is one responsibility, and knowing how to organize "work" is a different responsibility! And you can apply this on the function level, the component level (page layout vs product widget) and even on the system level. I work at Elastic and our main product Elasticsearch is a distributed database, so you have many nodes and each node has a different responsibility. So we have some data storage nodes and some search nodes. But we also have a node type that we call a coordinating node, it it's only job is to tell the other nodes how to do their jobs!
> 
> So to summarize, if you aim to create your code in smaller pieces that know as little as possible it'll be easier to maintain your code, and your folder structure should reflect this.
well said, It was exactly what I mean.
When I found an issue in the styling of one component then  I want to write a review but when I found there is only one CSS file for all the project codes.
I decided to start with this project structure suggestion and with this method, you will not need context anymore too.
 because I did the same and It will take a few days but afterwards it will be a good experience.

**Comment by islam-fawzy25** [comment]
> I have been thinking about this all day! And I realized that I did not address the main goal of Islam's comment! I should have asked "why was it best practice to split by component/container".
> 
> TL;DR: If you aim to create your code in smaller pieces that know as little as possible (single responsibility) it'll be easier to maintain your code, and your folder structure should reflect this.
> 
> Learning from the purpose of a best practice makes it so that we can apply the wisdom even if technology changes. In my view, the main goal of that pattern is the idea "single responsibility".
> 
> The main idea is that, the less things each part of your code knows about the better. Because the less things it knows about the less reason it has to change. And less reason for code to change means less work which also means less chances to put bugs into the code. If one change in your system requires that a lot of different pieces change then these pieces know too much about each other, and thus they likely have more than one responsibility.
> 
> So think of a simple React component that only accepts props and does some rendering. It knows very little, it's only responsible for rendering the data it receives to the screen, and maybe reacting to events by calling a callback it got from props. It knows nothing else, it only has to change if a) the rendering should change or b) the specific data it gets changed. For all other changes in the code base, it does not need to change. A container should be the opposite, it knows only about how to get the data and which component to pass that data to. It only needs to change if a) the data to fetch changes or b) there is another component to use for rendering.
> 
> So while we used to think of Pages as containers, there is actually two parts of a page. The data loading for that page, and the layout rendering for that page. So it would be better if that was two components, one that loads the data and one that only does the layout. And this is exactly what a hook is. A hook only handles data, it doesn't know about rendering. So you still gain the value of the single responsibility idea. A similar example, in your App you have all of the routes but it also contains other code. It would make sense to extract all of the routes to another component called AppRouter. That component only knows about how to define routes, and App only knows that there is some general layout and an AppRouter.
> 
> Related to this topic is the idea of "worker code" and "coordination code". If you imagine a recipe, it contains a set of steps to perform to achieve a tasty outcome. Each step doesn't really know about the other steps, and the steps could be combined in other ways to achieve a different (maybe) tasty outcome. So the thing that decides the outcome is the recipe itself, the thing that coordinates the steps. Code works in a similar way. You have functions that do "work" and then you have functions that only call other functions in the right order to get something done. And that's because knowing how to do "work" is one responsibility, and knowing how to organize "work" is a different responsibility! And you can apply this on the function level, the component level (page layout vs product widget) and even on the system level. I work at Elastic and our main product Elasticsearch is a distributed database, so you have many nodes and each node has a different responsibility. So we have some data storage nodes and some search nodes. But we also have a node type that we call a coordinating node, it it's only job is to tell the other nodes how to do their jobs!
> 
> So to summarize, if you aim to create your code in smaller pieces that know as little as possible it'll be easier to maintain your code, and your folder structure should reflect this.

well said, It was exactly what I mean.
When I found an issue in the styling of one component then  I want to write a review but when I found there is only one CSS file for all the project codes.
I decided to start with this project structure suggestion and with this method, you will not need context anymore too.
 because I did the same and It will take a few days but afterwards it will be a good experience.

**Comment by miltonhultgren** [comment]
I would recommend merging this PR and make follow up PRs that are more specific. 

So based on the feedback we've given so far, and your own ideas, create a few issues here in Github and then make PRs to submit smaller changes that are more specific!
